### DISCUSSION TITLE
Setting the Signin CallbackUrl

### DISCUSSION THREAD

**Discussion by chr-ge** [nextauthjs/next-auth#1003] [Help] [answered]
I'm a bit confused on how to redicrect to a page after signin (how to set the callbackUrl).
In my `[...nextauth]` file I have the redirect callback:
```js
redirect: async (url, baseUrl) => {
  return url.startsWith(baseUrl)
    ? Promise.resolve(url)
    : Promise.resolve(baseUrl)
},
```
When I click the **signin** button (signin from next-auth/client), I get redirected to:
`http://localhost:3000/auth/signin?callbackUrl=http://localhost:3000/`

How can I set the `callbackUrl` to be `http://localhost:3000/dashboard` for example so after **signin** the user is redirected to the `dashboard` page?

**Comment by jaaneh** [comment, marked as answer]
You can pass a `callbackUrl` to the `signIn()` method, and use it like this:
```js
<div onClick={() => signIn('twitter', { callbackUrl: 'http://localhost:3000/path/to/callback' })}>
```

You can also run custom checks on the login route and check for a session, then redirect if true. Example using `getServerSideProps` and passing providers to the signin component for creation of sign in buttons.
```js
export const getServerSideProps = async ({ req }) => {
  const _providers = await providers()
  const session = await getSession({ req })

  if (session) {
    return {
      props: {},
      redirect: {
        destination: '/',
        permanent: false
      }
    }
  }

  return {
    props: {
      providers: _providers // pass providers to component
    }
  }
}
```

**Comment by Robert-OP** [comment]
I tried using the `callbackUrl` using `next-auth: 3.4.1`, but I cannot get it to be passed in my `redirect_uri` when making the Azure AD B2C provider as related to this issue https://github.com/nextauthjs/next-auth/issues/591.
